src/main/java/net/boomerangplatform/migration/changesets/flow/FlowDatabaseChangeLog.java

```diff
@@ -1917,4 +1920,135 @@ public void updatingdefaultimage(MongoDatabase db) throws IOException {
     workers.put("config", configs);
     collection.replaceOne(eq("name", "Task Configuration"), workers);
   }
+  
+  /*
+   * Migration required for Flow v4
+   * See: https://github.com/boomerang-io/roadmap/issues/368 for the v3 to v4 comparison
+   */
+  @ChangeSet(order = "113", id = "113", author = "Tyson Lawrie")
+  public void v4MigrationTaskTemplates(MongoDatabase db) throws IOException {
+
+    MongoCollection<Document> taskTemplatesCollection =
+        db.getCollection(collectionPrefix + "task_templates");
+
+    final FindIterable<Document> taskTemplateEntities = taskTemplatesCollection.find();
+    for (final Document taskTemplateEntity : taskTemplateEntities) {
+      Document newTaskTemplateEntity = new Document();
+      newTaskTemplateEntity.put("name", taskTemplateEntity.get("name"));
```

Comment:
1. Instead of creating new documents for the collection, we can just modify the existing `taskTemplateEntity`
2. For properties that have no change, we would not need a `put` on the existing document.

Reply:
Hey @amhudson, is the creation of a new document an expensive operation?

As I either create a new document and add certain elements or delete elements from the old one, which is also slightly logically harder as we are flattening the structure.

Previously there was 1 task template holding many revisions. Now we have one template per revision(version).

Reply:
Ah cool. I missed what you were doing with the _id's there. Makes sense now.